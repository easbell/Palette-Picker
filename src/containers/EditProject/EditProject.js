import React from 'react'
import { connect } from 'react-redux'
import '../../components/DisplayProject/DisplayProject.css'
import EditPalette from '../EditPalette/EditPalette'
import { setProjects, setPalettes } from '../../actions'
import { deleteFromBE } from '../../helpers/apiCalls'
import { withRouter } from 'react-router-dom'

export const EditProject = (props) => {
  const { name, id } = props.foundProject
  const palettes = props.palettes.filter(palette => {
    return palette.project_id === id
  }).map(palette => {
      return <EditPalette key={palette.id} palette={palette}/>
    })

  const deleteProject = (id) => {
    const updatedProjects = props.projects.filter(project => {
      return project.id !== id
    })
    props.setProjects(updatedProjects)
    const url = process.env.REACT_APP_BACKEND_URL + `api/v1/projects/${id}`
    deleteFromBE(url, id)
    deleteLinkedPalettes(id)
  }

  const deleteLinkedPalettes = (id) => {
    props.palettes.filter(palette => {
      return palette.project_id === id
    }).forEach(palette => {
      const url = process.env.REACT_APP_BACKEND_URL + `api/v1/palettes/${palette.id}`
      deleteFromBE(url, palette.id)
    })
    const unlinkedPalettes = props.palettes.filter(palette => {
      return palette.project_id !== id
    })
    setPalettes(unlinkedPalettes)
    props.history.push('/my-projects')
  }

  return (
    <div className='project-card edit-project-card'>
      <h2 className='project-name'>{name}</h2>
      {palettes}
      <div className='delete-btn-container'>
        <button className='delete-project' onClick={() => deleteProject(id)}>Delete Project</button>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => ({
  palettes: state.palettes,
  projects: state.projects
})

const mapDispatchToProps = (dispatch) => ({
  setProjects: (projects) => dispatch(setProjects(projects)),
  setPalettes: (palettes) => dispatch(setPalettes(palettes))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(EditProject))