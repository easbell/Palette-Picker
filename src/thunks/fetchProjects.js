import { isLoading, setProjects, hasErrored } from '../actions'

export const fetchProjects = () => {
  const url = process.env.REACT_APP_BACKEND_URL + 'api/v1/projects';
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const projects = await response.json()
      dispatch(isLoading(false))
      dispatch(setProjects(projects))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}