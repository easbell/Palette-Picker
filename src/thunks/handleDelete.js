import { isLoading, hasErrored } from '../actions'

export const handleDelete = (url, action, items, id) => {
  return async (dispatch) => {
    dispatch(isLoading(true))
    try {
      const response = await fetch(url, {
        method: 'DELETE',
        body: JSON.stringify({id: id}),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      const msg = await response.json()
      console.log(msg)
      dispatch(isLoading(false))
      dispatch(action(items))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}