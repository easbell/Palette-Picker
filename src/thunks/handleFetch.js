import { isLoading, hasErrored } from '../actions'

export const handleFetch = (url, action, options) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, options)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      dispatch(isLoading(false))
      dispatch(action(data))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}

