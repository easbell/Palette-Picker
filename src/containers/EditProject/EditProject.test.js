import React from 'react';
import { shallow } from 'enzyme';
import { EditProject, mapDispatchToProps, mapStateToProps } from './EditProject';
import { handleFetch } from '../../thunks/handleFetch';
import { handleDelete } from '../../thunks/handleDelete';

describe('EditProject', () => {
  let wrapper;
  let mockPalettes;
  let mockEvent;
  let mockFn;

  beforeEach(() => {
    mockPalettes = [{ id: 1, project_id: 1}];
    mockFn = jest.fn();
    wrapper = shallow( 
      <EditProject 
        foundProject={{ name: 'name', id: 1 }}
        projects={[{name: 'name'}, {name: 'other'}]}
        palettes={mockPalettes}
        handleDelete={mockFn}
        handleFetch={mockFn}
        history={[]}/>
    )
    mockEvent = { target: {name: 'projectName', value: 'name' } }
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state upon handleChange', () => {
    wrapper.find('.project-name').simulate('change', mockEvent);
    expect(wrapper.state('projectName')).toBe('name');
  });

  it('should invoke updateProject upon submit', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateProject');
    wrapper.find('.edit-project-card').simulate('submit', {preventDefault: () => {} })
    expect(spy).toHaveBeenCalled();
  });

  it('should dispatch handleFetch when updateProject is invoked', async () => {
    await wrapper.instance().updateProject();
    expect(wrapper.instance().props.handleFetch).toHaveBeenCalled();
  });

  it('should dispatch handleDelete when deleteProject is invoked', async () => {
    await wrapper.instance().deleteProject()
    expect(wrapper.instance().props.handleDelete).toHaveBeenCalled();
  });
  
  describe('mapStateToProps', () => {
    it('returns correct props', () => {
      const mockState = {
        projects: ['projects'],
        palettes: ['palettes'],
        other: 'other'
      }
      const expectedProps = {
        projects: ['projects'],
        palettes: ['palettes']
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expectedProps)
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch with handleFetch action', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.handleFetch();
      mappedProps.handleDelete();
      expect(mockDispatch).toHaveBeenCalledTimes(2);
    });
  });
});