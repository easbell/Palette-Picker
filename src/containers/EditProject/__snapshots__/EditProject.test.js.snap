// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProject should match snapshot 1`] = `
<form
  className="project-card edit-project-card"
  onSubmit={[Function]}
>
  <input
    className="project-name"
    onChange={[Function]}
    value="name"
  />
  <ConnectFunction
    key="1"
    palette={
      Object {
        "id": 1,
        "project_id": 1,
      }
    }
  />
  <div
    className="delete-btn-container"
  >
    <button
      className="update-project"
      type="submit"
    >
      Save
    </button>
    <button
      className="update-project delete"
      onClick={[Function]}
    >
      Delete Project
    </button>
  </div>
</form>
`;
